### DISCUSSION TITLE
React 19 compiler + jotai

### DISCUSSION THREAD

**Discussion by lveillard** [pmndrs/jotai#2546] [General]
Since useMemo will "disappear" with the React Compiler, I wonder how would these jotai pattern evolve:
```ts
//atoms.ts
export const unitMapAtom = atom<Record<string, Unit>>({});
export const unitAtom = (id: string) => atom((get): Unit | undefined => get(unitMapAtom)[id]);
```
```ts
//MyComp.tsx

export const MyComp({unitId}) => 
   {
      const unit = useAtomValue(useMemo(() => unitAtom(unitId), [unitId]));
      return <div> {JSON.stringify(unit}</div>
   }
```
If the compiler does this kind of code automatically include useMemo, wouldnt that cause any problems to jotai's normal usage or maybe to patterns similar to this one?

📝 For context, this pattern enables to pick a particular element of a big map dynamically, when the parameter (in this example `unitId`) can't be stored in another atom. It's an alternative to atomFamilies.

📝 React 19.0.0 will not include the compiler, it will arrive later.

**Comment by dai-shi** [comment]
Anything is uncertain, but I feel like, a) if we keep `useMemo`, React Compiler won't do anything, and b) if we omit it, React Compiler may (or may not?) automatically add memo capabilities.

**Reply by dai-shi** [reply]
> useMemo will "disappear" with the React Compiler

I think I missed it.

[React Compiler Playground](https://playground.react.dev/#N4Igzg9grgTgxgUxALhAgHgBwjALgAgDMoA7OXASwhPwFkBPAYQgFtMAKYfUi3ASQAm+AL4BKfMAA6NfHGpgCPAgF5uYBAEFcrAGoBDADZQE7KOtoIWEdu3HKAfNxK8trU8-4DRAGnwBtJUEAXVFRAG5pfHwYBFxYGgAeAQoAN0dgACkAZQB5ADkAOgUYChIAcwpCendeUWEEgHpktOlhEGEgA)

Yeah, it disappears, but it doesn't look very bad.

**Reply by dai-shi** [reply]
If we have an issue that can't be solved, the worst case would be to provide `useAtomMemo` from Jotai which avoids the compilation.

**Comment by lveillard** [comment]
I just tested nextjs + react 19 with the experimental compiler. We have our our code full of atoms and useMemo() which purpose are keeping () => atom patterns from making our frontend re-render indefinitely.

The result: Infinite re-renders everywhere :S

**Reply by dai-shi** [reply]
Thanks for testing. Can you share a minimal reproduction?

**Reply by angelo-hub** [reply]
+1 too, we now feel trapped on jotai, when zustand fully supports the compiler since it's based on proxies

**Reply by dai-shi** [reply]
Can you create a minimal repro with https://stackblitz.com please?

**Reply by angelo-hub** [reply]
have you attempted it? just memoize an atom, to be fair we also use `useReducer` before you guys decided to deprecate it, it made a lot of sense for us for certain situations. 

I am still not fond of the deprecation of such a common method

**Reply by angelo-hub** [reply]
useReducer is not often used but when we're talking about login/onboarding actions for an enterprise app, it's parimount to have decisive actions that can be executed on a derived state object

**Reply by dai-shi** [reply]
> have you attempted it?

No I haven't. Have you?

> I am still not fond of the deprecation of such a common method

That happens. It's added a very long time ago when we weren't experienced in the API design.

> it made a lot of sense for us for certain situations.

Can you share such situations? If it makes sense more than just a recipe, we are happy to create a third-party library.

**Comment by redkean** [comment]
I just tested mine today and don't seem to see any issues. will need check my other implementations though.
